elide-core/src/main/java/com/yahoo/elide/security/PermissionManager.java

```diff
@@ -0,0 +1,278 @@
+        }
+    }
+
+    public void executeCommitChecks() {
+        commitChecks.forEach(Supplier::get);
+    }
+
+    private void runPermissionChecks(Class<? extends Check>[] checks,
+                                     boolean isAny,
+                                     PersistentResource<?> resource,
+                                     ChangeSpec<?> changeSpec,
+                                     Class<? extends Annotation> annotationClass,
+                                     boolean isFieldAware) {
+        if (resource.getRequestScope().getSecurityMode() == SecurityMode.BYPASS_SECURITY) {
+            return;
+        }
+
+        BiFunction<Check, Optional<ChangeSpec<?>>, Boolean> checkFn =
+                (check, changespec) -> check.ok(resource.getRequestScope(), changespec);
+
+        boolean shouldDefer = CreatePermission.class.equals(annotationClass)
+                || UpdatePermission.class.equals(annotationClass);
+
+        if (!shouldDefer) {
+            checkFn = (check, changespec) -> check.ok(resource.getRequestScope(), changespec)
+                    && check.ok(resource.getObject(), resource.getRequestScope(), changespec);
+        }
+
+        executePermissions(checks,
+                isAny ? CheckMode.ANY : CheckMode.ALL,
+                resource,
+                Optional.ofNullable(changeSpec),
+                shouldDefer,
+                isFieldAware,
+                checkFn);
+    }
+
+    /**
+     * Execute a set of permission checks.
+     *
+     * @param checks Array of Check annotations
+     * @param mode true if ANY, else ALL
+     * @param resource provided PersistentResource to check
+     * @param changeSpec specification of changes
+     * @param shouldDefer Determine whether or not this check should be re-run in its deferred form (i.e. commitCheck)
+     * @param isFieldAware Determine whether or not this is a field-aware check
+     * @param checkFn the function that actually executes the check so this logic can be reused
+     */
+    private void executePermissions(Class<? extends Check>[] checks,
+                                    CheckMode mode,
+                                    PersistentResource<?> resource,
+                                    Optional<ChangeSpec<?>> changeSpec,
+                                    boolean shouldDefer,
+                                    boolean isFieldAware,
+                                    BiFunction<Check, Optional<ChangeSpec<?>>, Boolean> checkFn) {
+        // TODO: Need to write field-aware checks.
+        for (Class<? extends Check> check : checks) {
+            Check checkHandler;
+            try {
+                checkHandler = check.newInstance();
+            } catch (InstantiationException | IllegalAccessException e) {
+                throw new InvalidSyntaxException("Illegal permission check " + check.getName(), e);
+            }
+
+            Boolean ok = checkFn.apply(checkHandler, changeSpec);
```

Comment:
This is not exactly straight-forward, I suppose.

The `Check` interface encapsulates 2 methods. One for `Commit` checks and one for `Operation` checks. The first time through, we *always* execute the operation check (which may be implemented as nothing but `true` to ignore its affects). The deferral variable determines whether or not we should queue up this check to execute its `commit` check method at a later time.

The reason for the `shouldDefer` in the first place is short-circuiting. If we always defer and we have an `any` check for a Read, for instance, then we would defer our decision until commit time even though we don't support commit checks for Reads.